File: kisan_warehouse/outwards/report/payment_pending_outwards/payment_pending_outwards.js
// Copyright (c) 2025, Kisan Warehouse and contributors
// For license information, please see license.txt

frappe.query_reports["Payment Pending Outwards"] = {
    "filters": [
        {
            "fieldname": "customer",
            "label": __("Customer"),
            "fieldtype": "Link",
            "options": "Customer",
            "width": "80px"
        },
        {
            "fieldname": "broker",
            "label": __("Broker"),
            "fieldtype": "Link",
            "options": "Broker",
            "width": "80px"
        },
        {
            "fieldname": "product",
            "label": __("Product"),
            "fieldtype": "Link",
            "options": "Product",
            "width": "80px"
        },
        {
            "fieldname": "warehouse",
            "label": __("Warehouse"),
            "fieldtype": "Link",
            "options": "Warehouse",
            "width": "80px"
        },
        {
            "fieldname": "payment_due_date_from",
            "label": __("Payment Due Date (From)"),
            "fieldtype": "Date",
            "width": "80px"
        },
        {
            "fieldname": "payment_due_date_to",
            "label": __("Payment Due Date (To)"),
            "fieldtype": "Date",
            "width": "80px"
        },
        {
            "fieldname": "payment_status",
            "label": __("Payment Status"),
            "fieldtype": "Select",
            "options": ["", "Pending", "Processing", "Success", "Failed"],
            "width": "80px"
        },
        {
            "fieldname": "show_all",
            "label": __("Show All (Ignore Date Filter)"),
            "fieldtype": "Check",
            "width": "80px"
        }
    ],

    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        
        if (column.fieldname == "sauda_id" && data) {
            if (data.sauda_id && data.sauda_id !== "-") {
                value = `<a href="/app/sauda/${data.sauda_id}" style="color: #007bff;">${data.sauda_id}</a>`;
            } else {
                value = `<span style="color: #6c757d; font-style: italic;">No Sauda</span>`;
            }
        }
        
        if (column.fieldname == "days_status" && data) {
            if (data.urgency == "Overdue") {
                value = `<span style="color: #dc3545; font-weight: bold;">${data.days_status}</span>`;
            } else if (data.urgency == "Due Today") {
                value = `<span style="color: #fd7e14; font-weight: bold;">${data.days_status}</span>`;
            } else {
                value = `<span style="color: #28a745; font-weight: 500;">${data.days_status}</span>`;
            }
        }
        
        if (column.fieldname == "total_amount_pending" && data) {
            if (data.total_amount_pending > 0) {
                if (data.urgency == "Overdue") {
                    value = `<span style="color: #dc3545; font-weight: 500;">₹${data.total_amount_pending}</span>`;
                } else if (data.urgency == "Due Today") {
                    value = `<span style="color: #fd7e14; font-weight: 500;">₹${data.total_amount_pending}</span>`;
                } else {
                    value = `<span style="color: #6c757d;">₹${data.total_amount_pending}</span>`;
                }
            }
        }
        
        if (column.fieldname == "payment_status" && data) {
            let statusColor = {
                'pending': '#fd7e14',
                'processing': '#17a2b8', 
                'success': '#28a745',
                'failed': '#dc3545'
            };
            let color = statusColor[data.payment_status] || '#6c757d';
            value = `<span style="color: ${color}; font-weight: 500;">${data.payment_status}</span>`;
        }

        if (column.fieldname == "outward_id" && data && data.urgency == "Overdue") {
            $(row).css('background-color', '#fff5f5');
        }
        
        return value;
    },

    "onload": function(report) {
        report.page.add_inner_button(__("🏦 Bank Export"), function() {
            let raw_data = frappe.query_report.data;
            
            let valid_data = raw_data.filter(row => {
                return row.outward_id && 
                       row.outward_id.startsWith('OUT-') &&
                       row.customer_name && 
                       row.customer_name !== 'Total' &&
                       row.total_amount_pending > 0;
            });
            
            if (!valid_data || valid_data.length === 0) {
                frappe.msgprint(__("No valid payment records to export. Please apply filters and ensure the report has data."));
                return;
            }
            
            export_to_bank_format(valid_data);
        });

        report.page.add_inner_button(__("Reset All Filters"), function() {
            frappe.query_report.set_filter_value("customer", "");
            frappe.query_report.set_filter_value("broker", "");
            frappe.query_report.set_filter_value("product", "");
            frappe.query_report.set_filter_value("warehouse", "");
            frappe.query_report.set_filter_value("payment_due_date_from", "");
            frappe.query_report.set_filter_value("payment_due_date_to", "");
            frappe.query_report.set_filter_value("payment_status", "");
            frappe.query_report.set_filter_value("show_all", 0);
            frappe.query_report.refresh();
        });

        report.page.add_inner_button(__("Today's Due"), function() {
            frappe.query_report.set_filter_value("payment_due_date_from", frappe.datetime.get_today());
            frappe.query_report.set_filter_value("payment_due_date_to", frappe.datetime.get_today());
            frappe.query_report.set_filter_value("show_all", 0);
            frappe.query_report.refresh();
        });

        report.page.add_inner_button(__("Overdue Payments"), function() {
            frappe.query_report.set_filter_value("payment_due_date_from", "");
            frappe.query_report.set_filter_value("payment_due_date_to", frappe.datetime.add_days(frappe.datetime.get_today(), -1));
            frappe.query_report.set_filter_value("show_all", 0);
            frappe.query_report.refresh();
        });

        report.page.add_inner_button(__("All Pending"), function() {
            frappe.query_report.set_filter_value("show_all", 1);
            frappe.query_report.set_filter_value("payment_due_date_from", "");
            frappe.query_report.set_filter_value("payment_due_date_to", "");
            frappe.query_report.set_filter_value("payment_status", "");
            frappe.query_report.refresh();
        });

        report.page.add_inner_button(__("Upcoming Payments"), function() {
            frappe.query_report.set_filter_value("payment_due_date_from", frappe.datetime.add_days(frappe.datetime.get_today(), 1));
            frappe.query_report.set_filter_value("payment_due_date_to", "");
            frappe.query_report.set_filter_value("show_all", 0);
            frappe.query_report.refresh();
        });
    }
};

function export_to_bank_format(data) {
    frappe.show_alert({
        message: __('Preparing bank export file...'),
        indicator: 'blue'
    }, 3);

    if (typeof XLSX === 'undefined') {
        loadScript('https://cdnjs.cloudflare.com/ajax/libs/xlsx/0.18.5/xlsx.full.min.js', function() {
            generate_bank_export(data);
        });
    } else {
        generate_bank_export(data);
    }
}

function loadScript(src, callback) {
    var script = document.createElement('script');
    script.src = src;
    script.onload = callback;
    script.onerror = function() {
        frappe.msgprint(__('Failed to load Excel library. Please check your internet connection.'));
    };
    document.head.appendChild(script);
}

function generate_bank_export(data) {
    try {
        frappe.call({
            method: "kisan_warehouse.outwards.report.payment_pending_outwards.payment_pending_outwards.get_customer_bank_details",
            args: {
                customer_ids: data.map(row => row.customer_id || row.customer_name).filter(Boolean)
            },
            callback: function(response) {
                let customer_bank_data = response.message || {};
                create_bank_excel(data, customer_bank_data);
            }
        });
        
    } catch (error) {
        frappe.show_alert({
            message: __(`Export failed: ${error.message}`),
            indicator: 'red'
        }, 5);
    }
}

function create_bank_excel(data, customer_bank_data) {
    try {
        let bank_data = data.map((row) => {
            let message_type = (row.total_amount_pending > 200000) ? "RTGS" : "NEFT";
            
            let current_date = new Date();
            let formatted_date = [
                String(current_date.getDate()).padStart(2, '0'),
                String(current_date.getMonth() + 1).padStart(2, '0'),
                current_date.getFullYear()
            ].join('/');
            
            let customer_key = row.customer_id || row.customer_name;
            let bank_details = customer_bank_data[customer_key] || {};
            
            return {
                "Client Code": "ABCD1234",
                "Customer Reference No": row.outward_id || "",
                "Debit Account No.": "1122334455556",
                "Transaction Type Code": "LBT",
                "Message Type": message_type,
                "Beneficiary ID": "",
                "Beneficiary Name": row.customer_name || "",
                "Beneficiary Account No.": bank_details.bank_account_no || "",
                "Beneficiary Bank Swift Code / IFSC Code": bank_details.ifsc_code || "",
                "Payment Amount": parseFloat(row.total_amount_pending) || 0,
                "Value Date": formatted_date,
                "Remarks": `Payment for Outward ${row.outward_id} - ${row.product_name || 'Product'}`
            };
        });

        let wb = XLSX.utils.book_new();
        let ws = XLSX.utils.json_to_sheet(bank_data);
        
        ws['!cols'] = [
            {wch: 12}, {wch: 22}, {wch: 18}, {wch: 22}, {wch: 15}, {wch: 15},
            {wch: 30}, {wch: 22}, {wch: 25}, {wch: 15}, {wch: 12}, {wch: 40}
        ];
        
        XLSX.utils.book_append_sheet(wb, ws, "Bank Export");
        
        let current_date = new Date();
        let date_str = [
            current_date.getFullYear(),
            String(current_date.getMonth() + 1).padStart(2, '0'),
            String(current_date.getDate()).padStart(2, '0')
        ].join('');
        let time_str = [
            String(current_date.getHours()).padStart(2, '0'),
            String(current_date.getMinutes()).padStart(2, '0')
        ].join('');
        
        let filename = `Bank_Export_Payment_Pending_Outwards_${date_str}_${time_str}.xlsx`;
        
        XLSX.writeFile(wb, filename);
        
        let total_amount = bank_data.reduce((sum, row) => sum + parseFloat(row["Payment Amount"]), 0);
        let rtgs_count = bank_data.filter(row => row["Message Type"] === "RTGS").length;
        let neft_count = bank_data.filter(row => row["Message Type"] === "NEFT").length;
        
        let missing_bank_details = bank_data.filter(row => 
            !row["Beneficiary Account No."] || 
            !row["Beneficiary Bank Swift Code / IFSC Code"]
        ).length;
        
        frappe.show_alert({
            message: __(`Bank export completed successfully!<br>
                        File: ${filename}<br>
                        Total Records: ${bank_data.length}<br>
                        Total Amount: ₹${total_amount.toLocaleString('en-IN')}<br>
                        RTGS: ${rtgs_count} | NEFT: ${neft_count}${missing_bank_details > 0 ? '<br><strong>⚠️ Warning: ' + missing_bank_details + ' records have missing bank details</strong>' : ''}`),
            indicator: 'green'
        }, 10);
        
    } catch (error) {
        frappe.show_alert({
            message: __(`Excel creation failed: ${error.message}`),
            indicator: 'red'
        }, 5);
    }
}

File: kisan_warehouse/outwards/report/payment_pending_outwards/payment_pending_outwards.py
# Copyright (c) 2025, Kisan Warehouse and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import getdate, today, formatdate, flt

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    
    return columns, data

def get_columns():
    """Define report columns with proper formatting"""
    return [
        {
            "label": _("Outward ID"),
            "fieldname": "outward_id",
            "fieldtype": "Link",
            "options": "Outward",
            "width": 140
        },
        {
            "label": _("Sauda ID"),
            "fieldname": "sauda_id",
            "fieldtype": "Link",
            "options": "Sauda",
            "width": 140
        },
        {
            "label": _("Customer"),
            "fieldname": "customer_name",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "label": _("Broker"),
            "fieldname": "broker_name",
            "fieldtype": "Data",
            "width": 120
        },
        {
            "label": _("Product"),
            "fieldname": "product_name",
            "fieldtype": "Data",
            "width": 120
        },
        {
            "label": _("Warehouse"),
            "fieldname": "warehouse_name",
            "fieldtype": "Data",
            "width": 130
        },
        {
            "label": _("Outward Date"),
            "fieldname": "outward_date",
            "fieldtype": "Date",
            "width": 100
        },
        {
            "label": _("Net Total Amount"),
            "fieldname": "net_total",
            "fieldtype": "Currency",
            "width": 130
        },
        {
            "label": _("Amount Paid"),
            "fieldname": "total_amount_paid",
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "label": _("Amount Pending"),
            "fieldname": "total_amount_pending",
            "fieldtype": "Currency",
            "width": 130
        },
        {
            "label": _("Payment Due Date"),
            "fieldname": "payment_due_date",
            "fieldtype": "Date",
            "width": 120
        },
        {
            "label": _("Payment Status"),
            "fieldname": "payment_status",
            "fieldtype": "Data",
            "width": 100
        },
        {
            "label": _("Days Status"),
            "fieldname": "days_status",
            "fieldtype": "Data",
            "width": 120
        }
    ]

def get_data(filters):
    """Fetch and process report data"""
    
    conditions = get_conditions(filters)
    
    query = """
        SELECT 
            o.name as outward_id,
            COALESCE(o.sauda, '') as sauda_id,
            COALESCE(CONCAT(c.first_name, ' ', c.last_name), 'No Customer') as customer_name,
            COALESCE(CONCAT(b.first_name, ' ', b.last_name), 'No Broker') as broker_name,
            COALESCE(p.product_name, 'No Product') as product_name,
            COALESCE(w.warehouse_name, 'No Warehouse') as warehouse_name,
            o.outward_date,
            COALESCE(o.net_total, 0) as net_total,
            COALESCE(o.total_amount_paid, 0) as total_amount_paid,
            COALESCE(o.total_amount_pending, o.net_total) as total_amount_pending,
            o.payment_due_date,
            COALESCE(o.payment_status, 'pending') as payment_status,
            DATEDIFF(CURDATE(), o.payment_due_date) as days_difference,
            o.customer as customer_id,
            COALESCE(c.bank_account_name, '') as bank_account_name,
            COALESCE(c.bank_account_no, '') as bank_account_no,
            COALESCE(c.ifsc_code, '') as ifsc_code,
            COALESCE(c.bank_name, '') as bank_name
        FROM `tabOutward` o
        LEFT JOIN `tabCustomer` c ON o.customer = c.name
        LEFT JOIN `tabBroker` b ON o.broker = b.name  
        LEFT JOIN `tabProduct` p ON o.product = p.name
        LEFT JOIN `tabWarehouse` w ON o.warehouse = w.name
        LEFT JOIN `tabSauda` s ON o.sauda = s.name
        WHERE o.docstatus < 2 
        AND s.booking_type = 'Outward / Sales' {conditions}
        ORDER BY o.payment_due_date ASC, o.total_amount_pending DESC
    """.format(conditions=conditions)
    
    data = frappe.db.sql(query, filters, as_dict=1)
    
    for row in data:
        if not row.total_amount_paid:
            row.total_amount_paid = 0
            
        if not row.total_amount_pending:
            row.total_amount_pending = row.net_total
            
        row.total_amount_pending = flt(row.net_total - row.total_amount_paid, 2)
        
        days_diff = row.days_difference
        if days_diff > 0:
            row.days_status = f"{days_diff} Days Overdue"
            row.urgency = "Overdue"
        elif days_diff == 0:
            row.days_status = "Due Today"
            row.urgency = "Due Today"
        else:
            days_future = abs(days_diff)
            row.days_status = f"Due in {days_future} Days"
            row.urgency = "Upcoming"
            
        if not row.sauda_id:
            row.sauda_id = "-"
            
        if not row.bank_account_name:
            row.bank_account_name = ""
        if not row.bank_account_no:
            row.bank_account_no = ""
        if not row.ifsc_code:
            row.ifsc_code = ""
        if not row.bank_name:
            row.bank_name = ""
    
    return data

def get_conditions(filters):
    """Build SQL WHERE conditions based on filters"""
    conditions = []
    
    conditions.append("o.payment_status IN ('pending', 'processing')")
    conditions.append("o.total_amount_pending > 0")
    
    if filters.get("customer"):
        conditions.append("o.customer = %(customer)s")
    
    if filters.get("broker"):
        conditions.append("o.broker = %(broker)s")
    
    if filters.get("product"):
        conditions.append("o.product = %(product)s")
    
    if filters.get("warehouse"):
        conditions.append("o.warehouse = %(warehouse)s")
    
    if filters.get("payment_due_date_from"):
        conditions.append("o.payment_due_date >= %(payment_due_date_from)s")
    
    if filters.get("payment_due_date_to"):
        conditions.append("o.payment_due_date <= %(payment_due_date_to)s")
    
    if filters.get("payment_status"):
        if filters.get("payment_status") == "Pending":
            conditions.append("o.payment_status = 'pending'")
        elif filters.get("payment_status") == "Processing":
            conditions.append("o.payment_status = 'processing'")
        elif filters.get("payment_status") == "Success":
            conditions.append("o.payment_status = 'success'")
        elif filters.get("payment_status") == "Failed":
            conditions.append("o.payment_status = 'failed'")
    
    if not filters.get("payment_due_date_from") and not filters.get("payment_due_date_to"):
        if not filters.get("show_all"):
            conditions.append("o.payment_due_date = CURDATE()")
    
    return " AND " + " AND ".join(conditions) if conditions else ""

@frappe.whitelist()
def get_customer_bank_details(customer_ids):
    """Get bank details for customers"""
    import json
    
    if isinstance(customer_ids, str):
        customer_ids = json.loads(customer_ids)
    
    if not customer_ids:
        return {}
    
    placeholders = ', '.join(['%s'] * len(customer_ids))
    
    query = f"""
        SELECT 
            name as customer_id,
            COALESCE(bank_account_name, '') as bank_account_name,
            COALESCE(bank_account_no, '') as bank_account_no,
            COALESCE(ifsc_code, '') as ifsc_code,
            COALESCE(bank_name, '') as bank_name
        FROM `tabCustomer` 
        WHERE name IN ({placeholders})
    """
    
    bank_details = frappe.db.sql(query, customer_ids, as_dict=True)
    
    result = {}
    for detail in bank_details:
        result[detail.customer_id] = {
            'bank_account_name': detail.bank_account_name,
            'bank_account_no': detail.bank_account_no,
            'ifsc_code': detail.ifsc_code,
            'bank_name': detail.bank_name
        }
    
    return result